clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java

```diff
@@ -3493,4 +3507,201 @@ void handleFailure(Throwable throwable) {
+                }
+            }
+        };
+    }
+
+    @Override
+    public ListOffsetsResult listOffsets(Map<TopicPartition, OffsetSpec> topicPartitionOffsets,
+                                         ListOffsetsOptions options) {
+
+        // preparing topics list for asking metadata about them
+        final Map<TopicPartition, KafkaFutureImpl<ListOffsetsResultInfo>> futures = new HashMap<>(topicPartitionOffsets.size());
+        final Set<String> topics = new HashSet<>();
+        for (TopicPartition topicPartition : topicPartitionOffsets.keySet()) {
+            topics.add(topicPartition.topic());
+            futures.put(topicPartition, new KafkaFutureImpl<>());
+        }
+
+        final long nowMetadata = time.milliseconds();
+        final long deadline = calcDeadlineMs(nowMetadata, options.timeoutMs());
+        // asking for topics metadata for getting partitions leaders
+        runnable.call(new Call("topicsMetadata", deadline, new LeastLoadedNodeProvider()) {
+
+            @Override
+            MetadataRequest.Builder createRequest(int timeoutMs) {
+                return new MetadataRequest.Builder(new MetadataRequestData()
+                    .setTopics(convertToMetadataRequestTopic(topics))
+                    .setAllowAutoTopicCreation(false));
+            }
+
+            @Override
+            void handleResponse(AbstractResponse abstractResponse) {
+                MetadataResponse response = (MetadataResponse) abstractResponse;
+
+                Map<String, Errors> errors = response.errors();
+                Cluster cluster = response.cluster();
+
+                // completing futures for topics with errors
+                for (Map.Entry<String, Errors> topicError : errors.entrySet()) {
+                    for (Map.Entry<TopicPartition, KafkaFutureImpl<ListOffsetsResultInfo>> future: futures.entrySet()) {
+                        if (future.getKey().topic().equals(topicError.getKey())) {
+                            future.getValue().completeExceptionally(topicError.getValue().exception());
+                        }
+                    }
+                }
+
+                // grouping topic partitions per leader
+                Map<Node, Map<TopicPartition, ListOffsetRequest.PartitionData>> leaders = new HashMap<>();
+                Map<TopicPartition, Optional<Integer>> leaderEpochs = new HashMap<>();
+                for (TopicMetadata topicMetadata : response.topicMetadata()) {
+                    for (PartitionMetadata partitionMetadata : topicMetadata.partitionMetadata()) {
+                        leaderEpochs.put(new TopicPartition(topicMetadata.topic(), partitionMetadata.partition()), partitionMetadata.leaderEpoch());
+                    }
+                }
+                for (Map.Entry<TopicPartition, OffsetSpec> entry: topicPartitionOffsets.entrySet()) {
+
+                    // avoid sending listOffsets request for topics with errors
```

Comment:
Not sure this makes sense. I'd expect we'd attempt to retry the metadata request if needed (e.g. to find the leader). I'm guessing this logic was copied from deleteRecords? We could possibly leave this as a follow-up and fix both APIs at the same time. If we don't fix it, it would be a regression in the consumer group tool because the consumer's list offset APIs have this retry logic.

Reply:
JIRA for `deleteRecords()`: https://issues.apache.org/jira/browse/KAFKA-8982